XenAdmin/Core/ExtensionMethods.cs

```diff
@@ -88,24 +90,79 @@ public static string Ellipsise(this string text, Rectangle rectangle, Font font)
+        {
+            var formattedValue = value;
+            if (!string.IsNullOrEmpty(value))
+            {
+                if (indent)
+                {
+                    formattedValue = PrependIndentation(formattedValue);
+                }
+                if (addTimestamp)
+                {
+                    formattedValue = PrependTimestamps(formattedValue);
+                }
+            }
+
+            builder.AppendLine(formattedValue);
+
+            if (addExtraLine)
+            {
+                builder.AppendLine();
+            }
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Prepend every line in the input value with the specified indentation level.
+        /// </summary>
+        /// <param name="value">The value to which indentation will be applied</param>
+        /// <param name="indent">The level of indentation, i.e. the number of spaces to prepend to every line in the value.</param>
+        /// <returns>The input value with prepended indentation./</returns>
+        private static string PrependIndentation(string value, int indent = DEFAULT_STRING_INDENTATION)
         {
-            var indentString = "";
-            var i = 0;
-            while (i++ < indent)
-                indentString += " ";
-            var newvalue = value.Replace(System.Environment.NewLine, string.Format("{0}{1}", System.Environment.NewLine, indentString));
-            return builder.Append(string.Format("{0}{1}", indentString, newvalue));
+            var indentString = new string(' ', indent);
+            var newValue = value.Replace(Environment.NewLine, $"{Environment.NewLine}{indentString}");
+            return $"{indentString}{newValue}";
         }
 
-        public static StringBuilder AppendIndented(this StringBuilder builder, StringBuilder value, int indent = 2)
+        /// <summary>
+        /// Prepend every line in the input value with a formatted string of <see cref="DateTime.Now"/>.
+        /// </summary>
+        /// <param name="value">The input value</param>
+        /// <param name="localize">true to format the string with the user's locale</param>
+        /// <returns>The input value with prepended timestamps/</returns>
+        public static string PrependTimestamps(string value, bool localize = true)
         {
-            var indentString = "";
-            var i = 0;
-            while (i++ < indent)
-                indentString += " ";
-            var newvalue = value.Replace(System.Environment.NewLine, string.Format("{0}{1}", System.Environment.NewLine, indentString));
-            return builder.Append(string.Format("{0}{1}", indentString, newvalue));
+            var timestamp = DateTime.Now;
+            var timestampString = HelpersGUI.DateTimeToString(timestamp, Messages.DATEFORMAT_HMS, localize);
+            // normalise all line endings before splitting

```

Comment:
- Is the normalisation necessary?
- I think a `:` or nothing at all is easier on the eye than `>`.
- We may want to consider adding the date too, because someone may have a big system and start the update on May 31, 10:00pm,  and by the time it finishes it  may be June 1, 3am.

Reply:
> Is the normalisation necessary?
No but I thought this might be used in the future for completely different purposes so I tried to keep it generalised. For instance, printing out a log returned by some linux stdout? It's a better safe than sorry approach.

> I think a : or nothing at all is easier on the eye than >.

Sure but I did not like it since it's already used in the string, as a separator for minutes/hours/seconds
```log
xrtmia-08-02:
1:51:54 PM:   'xrtmia-08-02' is entering maintenance mode... done.
1:52:24 PM:   Installing [XenServer product] 8.2.50 on 'xrtmia-08-02'...
```
```log
xrtmia-08-02:
1:51:54 PM>   'xrtmia-08-02' is entering maintenance mode... done.
1:52:24 PM>   Installing [XenServer product] 8.2.50 on 'xrtmia-08-02'...
```
```log
xrtmia-08-02:
1:51:54 PM |   'xrtmia-08-02' is entering maintenance mode... done.
1:52:24 PM |   Installing [XenServer product] 8.2.50 on 'xrtmia-08-02'...
```

> We may want to consider adding the date too, because someone may have a big system and start the update on May 31, 10:00pm, and by the time it finishes it may be June 1, 3am.

👍

Reply:
If the output is from linux, normalisation won't do anything because it won't contain the current environment's new line, but I suppose it would be good for mixed output. Ok, let's leave it in.

Happy to go with the pipe `|`?

Reply:
> Happy to go with the pipe |?
works for me 👍